content/blog/2020-12-16-compcert/index.md

```diff
@@ -0,0 +1,249 @@
+
+By composing these simulations, we get forward simulations from RTL to Asm, from Cminor to Asm, and from Cstrategy to Asm. 
+
+We can then derive backwards simulations for these three language pairs.
+
+The theorem of correctness then asserts that Csem is in backwards simulation with Cstrategy, and therefore in backwards simulation with Asm. 
+
+### What's Missing? 
+
+Short answer: a lot! CompCert doesn't say anything about the semantics of the preprocessed C code. Nothing about safety of the assembling or linking processes, either. It doesn't guarantee that the executable has no segfaults or null pointer exceptions. It doesn't assume compilation halts, since the OCaml and other unverified portions may not terminate. Neither the compiler nor the code it generates has promises for time or space efficiency. It can't reason about threaded programs. Whether due to CompCert's implementation or C's limitations, it's definitely fair to criticize CompCert for its lack of range. 
+
+## Beyond the Original CompCert Paper
+
+John Regehr and coauthors in a [PLDI 2011 paper][csmith] remarked
+
+> *“The striking thing about our CompCert results is that the middle end bugs we found in all other
+compilers are absent. As of early 2011, the under-development version of CompCert is the only
+compiler we have tested for which Csmith cannot find wrong-code errors. This is not for lack
+of trying: we have devoted about six CPU-years to the task. The apparent unbreakability of
+CompCert supports a strong argument that developing compiler optimizations within a proof
+framework, where safety checks are explicit and machine-checked, has tangible benefits for
+compiler users.”*
+
+The only bugs found in CompCert have been in the unverified frontend, which is now mostly verified, and in the assembly code generation, where a number was too large for a field in the instruction encoding. The latter error was not of concern to the authors as assemblers generally catch these errors, while optimization bugs are notoriously difficult to completely patch in unverified compilers.
+
+In 2015, static analysis tool vendor [AbsInt][absint] joined the project to market and maintain the software, while Xavier Leroy and INRIA continued to spearhead CompCert research efforts. This is why all of the CompCert code linked here is under the AbsInt organization. AbsInt's involvement with CompCert is alluring; a project where engineering issues are handled in parallel with research questions seems to promise faster progress.
+
+Airbus, an aerospace corporation, uses CompCert to write the software for its planes. A [2014 report][airbus] from the company indicated that the worst-case execution time of its programs had improved by 12%. The commercial success of CompCert speaks to its ability to hold its own as an optimizing compiler, although it still doesn't employ some more aggressive code optimizations. 
+
+### An Achilles' Heel: Separate Compilation
```

Comment:
I really like this section! Cool example from the SepCompCert paper.

Reply:
Thanks <3